src/storage/buffer_manager/bm_file_handle.cpp

```diff
@@ -62,124 +67,124 @@ void BMFileHandle::addNewPageGroupWithoutLock() {
-    std::shared_lock sLck{fhSharedMutex};
-    assert(pageIdx < numPages);
+    std::unique_lock xLck{fhSharedMutex};
+    assert(originalPageIdx < numPages);
     // Although getPageElementCursorForPos is written to get offsets of elements
     // in pages, it simply can be used to find the group/chunk and offset/pos in group/chunk for
     // any chunked data structure.
-    auto pageGroupIdxAndPosInGroup =
-        PageUtils::getPageElementCursorForPos(pageIdx, StorageConstants::PAGE_GROUP_SIZE);
-    // If we have not created a vector of locks and pageVersion array for each page in this
-    // group, first create them.
-    bool pageGroupLockAcquired = false;
-    while (!pageGroupLockAcquired) { // spinning wait
-        pageGroupLockAcquired = !pageGroupLocks[pageGroupIdxAndPosInGroup.pageIdx]->test_and_set();
-    }
-    if (pageVersions[pageGroupIdxAndPosInGroup.pageIdx].empty()) {
-        pageVersions[pageGroupIdxAndPosInGroup.pageIdx].resize(
-            StorageConstants::PAGE_GROUP_SIZE, UINT32_MAX);
+    auto [pageGroupIdx, _] =
+        PageUtils::getPageElementCursorForPos(originalPageIdx, StorageConstants::PAGE_GROUP_SIZE);
+    if (!walPageIdxGroups.contains(pageGroupIdx)) {
+        walPageIdxGroups.insert(std::make_pair(pageGroupIdx, std::make_unique<WALPageIdxGroup>()));
     }
-    pageGroupLocks[pageGroupIdxAndPosInGroup.pageIdx]->clear();
+    return pageGroupIdx;
 }
 
 void BMFileHandle::resetToZeroPagesAndPageCapacity() {
-    std::unique_lock xlock(fhSharedMutex);
-    numPages = 0;
-    pageCapacity = 0;
+    removePageIdxAndTruncateIfNecessary(0 /* pageIdx */);
     FileUtils::truncateFileToEmpty(fileInfo.get());
-    initPageStatesAndGroups();
 }
 
 void BMFileHandle::removePageIdxAndTruncateIfNecessary(common::page_idx_t pageIdx) {
     std::unique_lock xLck{fhSharedMutex};
-    removePageIdxAndTruncateIfNecessaryWithoutLock(pageIdx);
-}
-
-void BMFileHandle::removePageIdxAndTruncateIfNecessaryWithoutLock(
-    common::page_idx_t pageIdxToRemove) {
-    if (numPages <= pageIdxToRemove) {
+    if (numPages <= pageIdx) {
         return;
     }
-    for (auto pageIdx = pageIdxToRemove; pageIdx < numPages; ++pageIdx) {
-        pageStates[pageIdx].reset();
-    }
-    numPages = pageIdxToRemove;
+    numPages = pageIdx;
+    pageStates.resize(numPages);
     auto numPageGroups = getNumPageGroups();
     if (numPageGroups == frameGroupIdxes.size()) {
         return;
     }
     assert(numPageGroups < frameGroupIdxes.size());
     frameGroupIdxes.resize(numPageGroups);
-    pageGroupLocks.resize(numPageGroups);
-    pageVersions.resize(numPageGroups);
+    for (auto& [pageGroupIdx, _] : walPageIdxGroups) {
+        if (pageGroupIdx >= numPageGroups) {
+            walPageIdxGroups.erase(pageGroupIdx);
+        }
+    }
+    pageCapacity = numPageGroups * StorageConstants::PAGE_GROUP_SIZE;
+    // TODO(Reviewer): Should the physical file also be truncated here?
```

Comment:
I dont think so. `removePageIdxAndTruncateIfNecessary` should be called only when a transaction aborted but in that case we never actually write anything to the original file (we only may have written to the wal, which gets deleted anyways).